### PR TITLE
Animation Composition

### PR DESCRIPTION
**Pull request by james7132**
[RENDERED](https://github.com/james7132/rfcs/blob/animation-composition/rfcs/51-animation-composition.md)

This RFC is an extension of #49, detailing the low level details on how to compose the base primitives for more complex animations together via blending. This RFC stops short of the final step of applying the sampled values to the target animated properties, as that is separate and can be handled regardless of how the values are composed or sampled.

~~This will remain a draft until #49, or some variant of it that covers the same subject matter, gets merged.~~ This looks like it might be easier to implement than the actual storage solution.

### CODE REVIEW

**Comment by aunyks** [comment]
I’d love to have this feature so that I can blend multiple glTF animations.

**Comment by ValorZard** [comment]
Since Bevy doesn't currently have an editor, or any official debug UI for that matter, how would a user manipulate the animation state tree? In Unity and Godot, there is a fancy visual scripting editor that lets you build a animation graph quite easily, but I could see that becoming quite unwieldy in code. This wouldn't be a problem once we have an actual editor, but for the time being, managing an animation state machine could become quite troublesome.

**Comment by james7132** [comment]
> Since Bevy doesn't currently have an editor, or any official debug UI for that matter, how would a user manipulate the animation state tree? In Unity and Godot, there is a fancy visual scripting editor that lets you build a animation graph quite easily, but I could see that becoming quite unwieldy in code. This wouldn't be a problem once we have an actual editor, but for the time being, managing an animation state machine could become quite troublesome.

I agree here. While we're missing the editor tools for this, I would like to add some tooling that is similar to the current `AnimationPlayer` to make it easy to just use it as is, but ultimately this graph is supposed to be a low level primitive, not a high level interface.

**Review by cart — COMMENTED** [review]
I think this is the right general direction!

File: rfcs/51-animation-composition.md
Comment (james7132) [reply rfcs/51-animation-composition.md:416]: I think it's well inside the scope of this RFC. I think Unity and Unreal also have a similar feature for masking bones to apply a certain animation.

The solution to this would require us to do a per-bone weight blend during graph evaluation. Masked out bones would force the applied weight to zero. This can either be done with a bitmask from bone ID that is checked during blending.

File: rfcs/51-animation-composition.md
Comment (james7132) [reply rfcs/51-animation-composition.md:243]: > Can we instead utilize the DAG nature of the entity graph (and the fact that AnimationGraph operates on a child hierarchy of the DAG) to animate AnimationGraphs in parallel by starting at "the top" of the entity hierarchy (filtered down to entities with AnimationGraph) and "animating down"? Peer animation graphs could never reference the same entity. A given AnimationGraph should only need to mutate a given Bone entity's component once at the end of the graph execution.

This would require us to assume that graphs do not have overlapping trees. Consider the following hierarchy:

```
        A
      /     \
    B       C
           /    \
         D      E
```

If both A and C have animation graphs targeting C or D, they will still alias.

Unless we enforce that only one animation graph can work in a given hierarchy and/or that animation graphs only live on root entities, this likely is intractable.

File: rfcs/51-animation-composition.md
Comment (james7132) [reply rfcs/51-animation-composition.md:93]: The first common use case that I can think of is "any-time-after" transitions. Where a transition from a source animation is triggered automatically after a fixed time point has passed. That source animation will continue from the same time it had before, but the new animation it's transitioning to will start at zero. Both will advance at their own respective time scales while the transition blends smoothly from the source to the destination. While this is notably very low level, this can be used to build smooth transitions for a state machine like flow at a higher level.

It should be absolute time relative to the underlying clip. The delta is scaled based on the global, graph level, and clip level time scales.

File: rfcs/51-animation-composition.md
Comment (james7132) [reply rfcs/51-animation-composition.md:311]: Bikeshedding a bit here: `AnimatedEntityBindings` might be a better name.

The end goal is indeed reflect based "animate anything", and there's multiple in flight PRs to make that both possible and performant. This design is explicitly made to avoid making decisions that would block this end goal. Should this defer to #49 for more details on this or should this be elaborated in more detail the implementation *process* not just strategy? Since it clear we likely are going to go through the statically typed Transform based solution first which will be extended to support the "animate anything" use cases.

File: rfcs/51-animation-composition.md
Comment (cart) [reply rfcs/51-animation-composition.md:243]: > Unless we enforce that only one animation graph can work in a given hierarchy and/or that animation graphs only live on root entities, this likely is impossible to get it to parallelize without explicit deduplication.

Under these circumstances, I think it is safe to parallelize everything within a given depth of the hierarchy in parallel, as entities at a given depth are guaranteed to not contain each other in their DAG, which is the pattern I was trying to get at. In the graph above: B and C graphs could be animated in parallel.

Given that a graph will often be animating X "things" at once across many entities (bones, components, etc), and those things are parallelizable, and that we can parallelize graphs at the sibling level, it seems like that already introduces a reasonably high chance of core saturation, especially given the high likelihood of "sibling-ness" for animated things.

>  I also generally want us to avoid more full hierarchy traversals if we can.

Agreed. But when compared with the cost of constructing per-entity access lists each frame, I might still bet on simple hierarchy traversals winning the perf game.

File: rfcs/51-animation-composition.md
Comment (cart) [reply rfcs/51-animation-composition.md:311]: Yeah the challenge is that RFCs must be "implementable", meaning that logical dependencies in other RFCs should be merged first, and the path for implementation should be clear.

I think we have a number of paths forward, which all work for me:
1. Adapt this RFC to be more generic / use the generic design/terminology as defined in #49. Block merging on #49
2. Add the minimal amount of generic design / terminology to this RFC to unblock making it generic. #49 can move at its own pace and iterate on what is introduced here.
3. Adapt this RFC to be less generic and explicitly only target bone transform animation in a way that is implementable (probably just static transform types).

File: rfcs/51-animation-composition.md
Comment (cart) [reply rfcs/51-animation-composition.md:416]: Makes sense to me! I'm cool with adding it or punting it, as long as the scenario is acknowledged / accounted for in the RFC.

File: rfcs/51-animation-composition.md
Comment (cart) [reply rfcs/51-animation-composition.md:243]: >  B and C graphs could be animated in parallel

Important to also call out that B descendants with graphs and C descendants with graphs can also run in parallel with each other, regardless of depth (provided C descendants honor the rules outlined above).

Comment (cart) [reply rfcs/51-animation-composition.md:243]: The fact that A can't run in parallel with C doesn't seem like a huge loss under those circumstances. A less common pattern I think, given that graphs will generally be orchestrating distinct logical units.

File: rfcs/51-animation-composition.md
Comment (cart) [reply rfcs/51-animation-composition.md:93]: > For the propagation, this is largely just a QoL feature for letting you advance an entire subgraph at once.. 

Yeah I think the propagation part might be too "high level" and opinionated at this stage in the game. Useful for some scenarios and not others, "invalid" in some cases (scrubbing results in going "outside" the duration bounds of one animation but not the other, animations are configured to progress at different rates / times, etc). And the behavior (when needed) can be implemented reasonably easily using existing features (ex: `for_each_child_node(parent_node, |child| child.set_time(time)`)

Comment (cart) [reply rfcs/51-animation-composition.md:93]: (or `for_each_descendant_node`)

File: rfcs/51-animation-composition.md
Comment (james7132) [reply rfcs/51-animation-composition.md:243]: > I think it is safe to parallelize everything within a given depth of the hierarchy in parallel

I don't think we have a solid way of querying for all entities at a specific hierarchy level, unless we carefully do a full hierarchy breadth-first search, which I still think we should avoid doing single threaded. What we could do instad is just parallelize on all graphs regardless of level, but predicate any actual animation with an ancestor query to ensure there is no higher level graph that might alias against it. If most animation graphs are root or close to root level, these ancestor queries should be fairly cheap. However, this would force it to parallelize only at the graph level: the hierarchy below the graph and the sampling application for each graph must be single threaded. This means that any graph with a large animated hierarchy under it are inevitably going to be a straggler. If there is a way for us to do parallel deduplication without allocation, we can avoid these stragglers and ensure better work distribution across threads.

File: rfcs/51-animation-composition.md
Comment (james7132) [reply rfcs/51-animation-composition.md]: I went with option 3 here, since delivering composition is probably a stronger need right now than generic animation, especially without good editor support.

File: rfcs/51-animation-composition.md
Comment (james7132) [reply rfcs/51-animation-composition.md]: Swapped up how we're supposed to do parallelization, which maps to the current parallelization strategy introduced in https://github.com/bevyengine/bevy/pull/6785.

File: rfcs/51-animation-composition.md
Comment (cart) [reply rfcs/51-animation-composition.md]: That makes sense to me!

File: rfcs/51-animation-composition.md
Comment (nicopap) [reply rfcs/51-animation-composition.md:93]: When reading the preceding paragraph, I had this question: "This seems to try to enable composition, but why a graph system would improve on a simple weight-based approach?"

And this sentence comes in. It gives a feature that is impossible to accomplish without graphs: namely: putting a bunch of animations together and having the capacity to control them exactly like if they were only a single animation.

In this context, I think having time control on blend nodes is really useful.

File: rfcs/51-animation-composition.md
Comment (nicopap) [inline rfcs/51-animation-composition.md:220]: The animation system animates more than "bones", already with morph targets https://github.com/bevyengine/bevy/pull/8158 animations are not limited to skinned mesh bones. I'd like a more general wording.

File: rfcs/51-animation-composition.md
Comment (nicopap) [inline rfcs/51-animation-composition.md:159]: I would suggest adding a "this is explained in the 'Bindings' section". I didn't understand what "map of property paths to `Track`" meant until I read it (ie: the `GraphBindings` struct)

File: rfcs/51-animation-composition.md
Comment (nicopap) [inline rfcs/51-animation-composition.md:150]: I don't understand what properties a `Node` stores. A list of dependencies and weights?

File: rfcs/51-animation-composition.md
Comment (nicopap) [inline rfcs/51-animation-composition.md:208]: what does "active inputs" mean in this context? Are those the `weight` field of `Influence`s?

File: rfcs/51-animation-composition.md
Comment (nicopap) [inline rfcs/51-animation-composition.md:191]: > Indexed by BoneId

Is this true? The paragraph that directly follows say

> This is effectively a `Vec<f32>` indexed on `ClipId` mapping clips and their respective *cumulative* weights.

File: rfcs/51-animation-composition.md
Comment (nicopap) [inline rfcs/51-animation-composition.md:376]: The current system has a path cache right? Why would that be impossible to establish in the graph-based system? How is maintaining the cache less of a nightmare? It should have the same limitations.

File: rfcs/51-animation-composition.md
Comment (james7132) [reply rfcs/51-animation-composition.md]: Yep. I removed the mention of this.

File: rfcs/51-animation-composition.md
Comment (james7132) [reply rfcs/51-animation-composition.md:376]: The path cache is just a cache. It's not treated as a source of truth, just a opportunistic way to reduce  expensive searching. The binding mentioned here is more akin to relations and would have to be treated as a source of truth to get the full benefits, and that has the same mire of edge cases are still being worked out for the hierarchy.

File: rfcs/51-animation-composition.md
Comment (mockersf) [reply rfcs/51-animation-composition.md:150]: a node is either a blend node or a clip node. they would have dependencies, time setting, a clip

File: rfcs/51-animation-composition.md
Comment (mockersf) [inline rfcs/51-animation-composition.md:206]: > the final weight for the clip is the sum of all path's weights.

Shouldn't it be the mean? Is there a risk of a weight greater than 1?

**Comment by mockersf** [comment]
is https://github.com/bevyengine/bevy/pull/7130 usable for this graph?

**Comment by mockersf** [comment]
@cart are you OK to move forward with this RFC? In my opinion there are a few minor questions, but they should be answered when going for the implementation

**Review by cart — APPROVED** [review]
I'm happy with where this is at. I think this is a solid first step for animation composition!

File: rfcs/51-animation-composition.md
Comment (james7132) [reply rfcs/51-animation-composition.md:206]: I think it's not well mentioned here, but the final results are after normalizing the entire weight vector so that the sum of all weights is 1.0, even if a final weight is over 1.0, it's influence will still be proportional to the rest of the clips that are in the graph.

**Comment by james7132** [comment]
I know this is already merged, but I wanted to close out a few last loose ends.

> is [bevyengine/bevy#7130](https://github.com/bevyengine/bevy/pull/7130) usable for this graph?

I haven't fully reviewed that PR yet, but my impression is no. The construction I had in mind does not involve many HashMap lookups of any kind and the traversal really only needed jumping around a few items in a `Vec<Node>`, which should be exceptionally fast to evaluate multiple times per game tick. Though I don't think this would preclude us from including this variant of a graph in `bevy_graph` though.

**Pull request merged**
